// hermes-transform/dist/transform/Errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnexpectedTransformationState = exports.NodeIsMutatedError = exports.NodeIsDeletedError = exports.InvalidStatementError = exports.InvalidReplacementError = exports.InvalidRemovalError = exports.InvalidInsertionError = void 0;

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
class NodeIsDeletedError extends Error {}

exports.NodeIsDeletedError = NodeIsDeletedError;

class NodeIsMutatedError extends Error {}

exports.NodeIsMutatedError = NodeIsMutatedError;

class InvalidInsertionError extends Error {}

exports.InvalidInsertionError = InvalidInsertionError;

class UnexpectedTransformationState extends Error {}

exports.UnexpectedTransformationState = UnexpectedTransformationState;

class InvalidStatementError extends Error {}

exports.InvalidStatementError = InvalidStatementError;

class InvalidRemovalError extends Error {}

exports.InvalidRemovalError = InvalidRemovalError;

class InvalidReplacementError extends Error {}

exports.InvalidReplacementError = InvalidReplacementError;
