// hermes-transform/dist/generated/special-case-node-types/DeclareExportDeclaration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeclareExportDeclarationNamedWithDeclaration = DeclareExportDeclarationNamedWithDeclaration;
exports.DeclareExportDeclarationNamedWithSpecifiers = DeclareExportDeclarationNamedWithSpecifiers;
exports.DeclareExportDefaultDeclaration = DeclareExportDefaultDeclaration;

var _detachedNode = require("../../detachedNode");

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
function DeclareExportDefaultDeclaration(props) {
  const node = (0, _detachedNode.detachedProps)(props.parent, {
    type: 'DeclareExportDeclaration',
    declaration: (0, _detachedNode.asDetachedNode)(props.declaration),
    specifiers: [],
    source: null,
    default: true
  });
  (0, _detachedNode.setParentPointersInDirectChildren)(node);
  return node;
}

function DeclareExportDeclarationNamedWithDeclaration(props) {
  const node = (0, _detachedNode.detachedProps)(props.parent, {
    type: 'DeclareExportDeclaration',
    declaration: (0, _detachedNode.asDetachedNode)(props.declaration),
    specifiers: [],
    source: null,
    default: false
  });
  (0, _detachedNode.setParentPointersInDirectChildren)(node);
  return node;
}

function DeclareExportDeclarationNamedWithSpecifiers(props) {
  const node = (0, _detachedNode.detachedProps)(props.parent, {
    type: 'DeclareExportDeclaration',
    declaration: null,
    specifiers: props.specifiers.map(n => (0, _detachedNode.asDetachedNode)(n)),
    source: (0, _detachedNode.asDetachedNode)(props.source),
    default: false
  });
  (0, _detachedNode.setParentPointersInDirectChildren)(node);
  return node;
}
