// hermes-transform/dist/generated/special-case-node-types/ExportNamedDeclaration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExportNamedDeclarationWithDeclaration = ExportNamedDeclarationWithDeclaration;
exports.ExportNamedDeclarationWithSpecifiers = ExportNamedDeclarationWithSpecifiers;

var _detachedNode = require("../../detachedNode");

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
function ExportNamedDeclarationWithDeclaration(props) {
  const node = (0, _detachedNode.detachedProps)(props.parent, {
    type: 'ExportNamedDeclaration',
    declaration: (0, _detachedNode.asDetachedNode)(props.declaration),
    specifiers: [],
    source: null,
    exportKind: props.exportKind
  });
  (0, _detachedNode.setParentPointersInDirectChildren)(node);
  return node;
}

function ExportNamedDeclarationWithSpecifiers(props) {
  const node = (0, _detachedNode.detachedProps)(props.parent, {
    type: 'ExportNamedDeclaration',
    declaration: null,
    specifiers: props.specifiers.map(n => (0, _detachedNode.asDetachedNode)(n)),
    // $FlowFixMe[incompatible-call]
    source: (0, _detachedNode.asDetachedNode)(props.source),
    exportKind: props.exportKind
  });
  (0, _detachedNode.setParentPointersInDirectChildren)(node);
  return node;
}
